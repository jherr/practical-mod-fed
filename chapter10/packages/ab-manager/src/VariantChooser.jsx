import React from "react";
import variants from "ab-manager/variants";

const chooseVariant = (test) =>
  variants[test][Math.floor(Math.random() * variants[test].length)];

const VariantChooser = ({ test, variant, variations, ...props }) => {
  const [testVariant] = React.useState(variant || chooseVariant(test));
  const Component = variations[testVariant];
  return (
    <React.Suspense fallback={<div>Loading variant</div>}>
      <Component {...props} />
    </React.Suspense>
  );
};

export default VariantChooser;
