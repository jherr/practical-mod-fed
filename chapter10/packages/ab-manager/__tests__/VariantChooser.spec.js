import React from "react";
import renderer from "react-test-renderer";
import VariantChooser from "../src/VariantChooser";

it("Should pick variant A", () => {
  const VariantA = () => <div>A</div>;
  const tree = renderer
    .create(
      <VariantChooser
        variations={{
          a: VariantA,
        }}
        variant="a"
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("Should pick variant A", () => {
  const VariantB = () => <div>B</div>;
  const tree = renderer
    .create(
      <VariantChooser
        variations={{
          b: VariantB,
        }}
        variant="b"
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
