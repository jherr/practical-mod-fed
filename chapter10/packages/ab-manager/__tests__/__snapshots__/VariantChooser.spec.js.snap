// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should pick variant A 1`] = `
<div>
  A
</div>
`;

exports[`Should pick variant A 2`] = `
<div>
  B
</div>
`;
